File: examples/python/multi_timeframe.py
"""Compute indicators on multiple timeframes from a single 1-minute feed.

Wickra exposes resampling on the Rust side (in `wickra-data`); from Python we
roll up bars manually using NumPy because most users already have their data
in a NumPy array and don't need the streaming-resample infrastructure for
offline analysis.

Run with::

    python -m examples.python.multi_timeframe path/to/1m.csv
"""

from __future__ import annotations

import argparse
import csv
from typing import Iterable, List, Tuple

import numpy as np

import wickra as ta


# Columns the OHLCV layout requires; the CSV header must name every one.
REQUIRED_COLUMNS = ("timestamp", "open", "high", "low", "close", "volume")


def read_csv(path: str) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
    """Read an OHLCV CSV into typed columns.

    Raises:
        ValueError: if the file has no header, is missing a required column,
            holds no data rows, or contains a non-numeric value (the message
            names the offending row).
    """
    ts, o, h, l, c, v = [], [], [], [], [], []
    with open(path, newline="") as f:
        reader = csv.DictReader(f)
        if reader.fieldnames is None:
            raise ValueError(f"{path}: CSV has no header row")
        missing = [col for col in REQUIRED_COLUMNS if col not in reader.fieldnames]
        if missing:
            raise ValueError(
                f"{path}: CSV is missing required column(s): "
                f"{', '.join(missing)}; found: {', '.join(reader.fieldnames)}"
            )
        for line_no, row in enumerate(reader, start=2):
            try:
                ts.append(int(row["timestamp"]))
                o.append(float(row["open"]))
                h.append(float(row["high"]))
                l.append(float(row["low"]))
                c.append(float(row["close"]))
                v.append(float(row["volume"]))
            except (TypeError, ValueError) as exc:
                raise ValueError(
                    f"{path}: row {line_no} has a non-numeric value: {exc}"
                ) from exc
    if not ts:
        raise ValueError(f"{path}: CSV has a header but no data rows")
    return (
        np.asarray(ts, dtype=np.int64),
        np.asarray(o, dtype=np.float64),
        np.asarray(h, dtype=np.float64),
        np.asarray(l, dtype=np.float64),
        np.asarray(c, dtype=np.float64),
        np.asarray(v, dtype=np.float64),
    )


def resample(
    ts: np.ndarray,
    o: np.ndarray,
    h: np.ndarray,
    l: np.ndarray,
    c: np.ndarray,
    v: np.ndarray,
    bucket: int,
) -> Tuple[np.ndarray, ...]:
    """Aggregate bar-level columns into coarser buckets of size `bucket`.

    Raises:
        ValueError: if the input series is empty.
    """
    if ts.size == 0:
        raise ValueError("resample received an empty series")
    bucket_index = (ts // bucket).astype(np.int64)
    boundaries = np.diff(bucket_index, prepend=bucket_index[0] - 1) != 0
    group_ids = np.cumsum(boundaries) - 1
    n_groups = int(group_ids.max()) + 1

    new_ts = np.empty(n_groups, dtype=np.int64)
    new_o = np.empty(n_groups, dtype=np.float64)
    new_h = np.empty(n_groups, dtype=np.float64)
    new_l = np.empty(n_groups, dtype=np.float64)
    new_c = np.empty(n_groups, dtype=np.float64)
    new_v = np.empty(n_groups, dtype=np.float64)

    for g in range(n_groups):
        mask = group_ids == g
        new_ts[g] = ts[mask][0]
        new_o[g] = o[mask][0]
        new_h[g] = h[mask].max()
        new_l[g] = l[mask].min()
        new_c[g] = c[mask][-1]
        new_v[g] = v[mask].sum()
    return new_ts, new_o, new_h, new_l, new_c, new_v


def summarize(label: str, close: np.ndarray, high: np.ndarray, low: np.ndarray) -> None:
    if close.size == 0:
        print(f"  {label:<10} (empty series — nothing to summarize)")
        return
    rsi = ta.RSI(14).batch(close)
    macd = ta.MACD().batch(close)
    adx = ta.ADX(14).batch(high, low, close)
    last_macd_hist = macd[~np.isnan(macd[:, 2])][-1, 2] if np.any(~np.isnan(macd[:, 2])) else float("nan")
    last_adx = adx[~np.isnan(adx[:, 2])][-1, 2] if np.any(~np.isnan(adx[:, 2])) else float("nan")
    valid_rsi = rsi[~np.isnan(rsi)]
    last_rsi = valid_rsi[-1] if valid_rsi.size else float("nan")
    print(
        f"  {label:<10} bars={close.size:>5}  "
        f"last_close={close[-1]:>10.4f}  rsi={last_rsi:>6.2f}  "
        f"macd_hist={last_macd_hist:+.4f}  adx={last_adx:>6.2f}"
    )


def main() -> int:
    p = argparse.ArgumentParser(description=__doc__.splitlines()[0] if __doc__ else None)
    p.add_argument("path", help="Path to a 1-minute OHLCV CSV (timestamps in milliseconds)")
    args = p.parse_args()

    ts, o, h, l, c, v = read_csv(args.path)
    one_minute = 60_000

    print(f"Multi-timeframe view of {args.path}")
    summarize("1m", c, h, l)
    for label, bucket in [("5m", 5), ("15m", 15), ("1h", 60), ("4h", 240), ("1d", 1440)]:
        if ts.size < bucket:
            continue
        rs_ts, rs_o, rs_h, rs_l, rs_c, rs_v = resample(ts, o, h, l, c, v, bucket * one_minute)
        summarize(label, rs_c, rs_h, rs_l)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: examples/node/multi_timeframe.js
// Multi-timeframe indicators with the Wickra Node binding.
//
// Reads the bundled 1-minute BTCUSDT CSV (or a path passed on the command
// line), rolls it up to coarser timeframes (5m / 15m / 1h / 4h / 1d) and
// prints the last RSI(14), MACD(12,26,9) histogram and ADX(14) on each.
// Mirrors examples/python/multi_timeframe.py — wickra-data's Resampler is
// only exposed in Rust today, so the roll-up is computed inline here.
//
// Run with:
//   node examples/node/multi_timeframe.js [path/to/1m.csv]

const fs = require('node:fs');
const path = require('node:path');

const wickra = require('wickra');

const REQUIRED_COLUMNS = ['timestamp', 'open', 'high', 'low', 'close', 'volume'];
const DEFAULT_CSV = path.join(__dirname, '..', 'data', 'btcusdt-1m.csv');
const ONE_MINUTE_MS = 60_000;

function readCsv(csvPath) {
  const text = fs.readFileSync(csvPath, 'utf8');
  const lines = text.split(/\r?\n/).filter((line) => line.length > 0);
  if (lines.length === 0) {
    throw new Error(`${csvPath}: file is empty`);
  }
  const header = lines[0].split(',').map((cell) => cell.trim());
  const missing = REQUIRED_COLUMNS.filter((col) => !header.includes(col));
  if (missing.length > 0) {
    throw new Error(
      `${csvPath}: missing required column(s): ${missing.join(', ')}; found: ${header.join(', ')}`,
    );
  }
  if (lines.length === 1) {
    throw new Error(`${csvPath}: CSV has a header but no data rows`);
  }
  const idx = {};
  for (const col of REQUIRED_COLUMNS) {
    idx[col] = header.indexOf(col);
  }
  const cols = { timestamp: [], open: [], high: [], low: [], close: [], volume: [] };
  for (let i = 1; i < lines.length; i++) {
    const cells = lines[i].split(',');
    for (const col of REQUIRED_COLUMNS) {
      const value = Number(cells[idx[col]]);
      if (!Number.isFinite(value)) {
        throw new Error(
          `${csvPath}: row ${i + 1} column '${col}' is not numeric: ${JSON.stringify(cells[idx[col]])}`,
        );
      }
      cols[col].push(value);
    }
  }
  return cols;
}

// Roll an OHLCV series up to `bucketMs`-sized buckets keyed on each bar's
// `floor(timestamp / bucketMs)`. Input timestamps must be monotonic
// non-decreasing (the bundled BTCUSDT-1m dataset is contiguous, so this
// holds by construction).
function resample(cols, bucketMs) {
  if (cols.timestamp.length === 0) {
    throw new Error('resample: empty input series');
  }
  const out = { timestamp: [], open: [], high: [], low: [], close: [], volume: [] };
  let bucketStart = Math.floor(cols.timestamp[0] / bucketMs) * bucketMs;
  let [o, h, l, c, v] = [cols.open[0], cols.high[0], cols.low[0], cols.close[0], cols.volume[0]];

  for (let i = 1; i < cols.timestamp.length; i++) {
    const start = Math.floor(cols.timestamp[i] / bucketMs) * bucketMs;
    if (start === bucketStart) {
      if (cols.high[i] > h) h = cols.high[i];
      if (cols.low[i] < l) l = cols.low[i];
      c = cols.close[i];
      v += cols.volume[i];
    } else {
      out.timestamp.push(bucketStart);
      out.open.push(o);
      out.high.push(h);
      out.low.push(l);
      out.close.push(c);
      out.volume.push(v);
      bucketStart = start;
      o = cols.open[i];
      h = cols.high[i];
      l = cols.low[i];
      c = cols.close[i];
      v = cols.volume[i];
    }
  }
  // Flush the final open bucket.
  out.timestamp.push(bucketStart);
  out.open.push(o);
  out.high.push(h);
  out.low.push(l);
  out.close.push(c);
  out.volume.push(v);
  return out;
}

function summarize(label, cols) {
  if (cols.close.length === 0) {
    console.log(`  ${label.padEnd(5)} (empty)`);
    return;
  }
  const rsi = new wickra.RSI(14);
  const macd = new wickra.MACD(12, 26, 9);
  const adx = new wickra.ADX(14);

  let lastRsi = null;
  let lastHist = null;
  let lastAdx = null;
  for (let i = 0; i < cols.close.length; i++) {
    const r = rsi.update(cols.close[i]);
    if (r !== null) lastRsi = r;
    const m = macd.update(cols.close[i]);
    if (m) lastHist = m.histogram;
    const a = adx.update(cols.high[i], cols.low[i], cols.close[i]);
    if (a) lastAdx = a.adx;
  }
  const lastClose = cols.close[cols.close.length - 1];
  const fmtR = lastRsi === null ? '    --' : lastRsi.toFixed(2).padStart(6);
  const fmtH = lastHist === null ? ' --   ' : `${lastHist >= 0 ? '+' : ''}${lastHist.toFixed(2)}`.padStart(6);
  const fmtA = lastAdx === null ? '    --' : lastAdx.toFixed(2).padStart(6);
  console.log(
    `  ${label.padEnd(5)} bars=${String(cols.close.length).padStart(5)}  ` +
      `last_close=${lastClose.toFixed(2).padStart(10)}  ` +
      `rsi=${fmtR}  macd_hist=${fmtH}  adx=${fmtA}`,
  );
}

function main() {
  const csvPath = process.argv[2] || DEFAULT_CSV;
  let cols;
  try {
    cols = readCsv(csvPath);
  } catch (err) {
    console.error(`error: ${err.message}`);
    process.exit(1);
  }
  console.log(`Multi-timeframe view of ${csvPath}`);
  summarize('1m', cols);
  for (const [label, mins] of [
    ['5m', 5],
    ['15m', 15],
    ['1h', 60],
    ['4h', 240],
    ['1d', 1440],
  ]) {
    summarize(label, resample(cols, mins * ONE_MINUTE_MS));
  }
}

main();
